renderdoc/serialise/codecs/vk_cpp_codec/ext_object.h

```diff
@@ -0,0 +1,304 @@
+/******************************************************************************
+* The MIT License (MIT)
+*
+* Copyright (c) 2018 Google LLC
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to deal
+* in the Software without restriction, including without limitation the rights
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in
+* all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+* THE SOFTWARE.
+******************************************************************************/
+#pragma once
+
+#include <assert.h>
+#include <math.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <array>
+#include <list>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "common/common.h"
+#include "core/core.h"
+#include "driver/vulkan/vk_common.h"
+#include "driver/vulkan/vk_resources.h"
```

Comment:
I think the ideal way to resolve this is to use the split between generic/vulkan-specific code I mentioned before. There's no particularly special reason a codec must exist only in `renderdoc/serialise/codecs` it's just that the previous ones were all agnostic and could be placed in the generic code.

If we had the general C++ export non-vulkan specific code in `renderdoc/serialise/codecs/cpp_export/` and then had `renderdoc/driver/vulkan/cpp_export/` with vulkan-specific code re-using the driver and actually declaring the vulkan version of the export, and referencing the generic C++ export code as needed.

It's not clear yet whether this will be a separate exporter per API with some way to selectively disable them depending on what is loaded, or a single entry with a dispatch to the right API, but that can be decided once things are further along.

Reply:
No worries, I'm hopeful that there will be a decent amount to extract out but I don't want to burden your initial PRs with making that distinction. Anything trivially separable would be nice but otherwise you don't have to worry about it, it will be on whoever does the implementation for the next API to extract out useful shared features.

Any code that includes any headers from the vulkan driver though should be within the vulkan driver folder, otherwise you risk breaking the build if e.g. the vulkan driver is disabled and excluded from the build.

Reply:
I've addressed this in current PR, so I'll resolve this comment. In the next PR's I'll change the cpp codec structure to respect the existing driver structure.